### ISSUE TITLE
Improve accessibility support

### ISSUE THREAD

**Issue by funkenstrahlen** [LeoNatan/LNPopupController#64]
I would like to be able to set some accessibility variables for `popupItem.title`, `popupItem.subtitle`, the and whole minified `popupBar` and the `progressBar` of the minified popup.

To do this I need access to the title and subtitle UILabels.

Also I need to access the popupbar of the minified player somehow. I do not know which object this is. I want to make it look like a button for a voiceover user with the hint that tapping it maxes the popup.

I also have the problem that when the popup is maxed elements underneath the top coverart UIImageView get still recognized as voiceover objects even if they are not visible to the user.

**Comment by LeoNatan** [comment]
Please explain what you need exactly. I am not interested at this point exposing the labels. How does it work with navigation bars, which also do not expose the label?

This all should be done at the framework level. So please make a list of the changes required and I will try to incorporate them. For example, one of them being the popup bar's button hint.

Regarding the last sentence, I am not sure, and need to investigate.

**Comment by funkenstrahlen** [comment]
I see I have not been very clear. I want to do 3 things which currently do not work:

* change the accessibility attributes of the dismiss button when the popup is maxed.
* if the popup is maxed, elements of the view the popup overlays "shine through" in voiceover mode. In my case a UITableView. this breaks a lot of things. I do not know how to fix that. I discovered it only happens in the top area of my popup view. In the bottom area I use an UIEffectsView. Maybe that blocks the underlaying view? Don't know...
* add an indication for the voiceover user that tapping the minified popup bar is possible. So make it appear like a button + hint. this should be added to the title + subtitle area in the middle as the barbuttonItems on the sides have other functionality.

The accessibility attributes need be publicly accessibly for me, because I want to add i18n support for VoiceOver and you might not want to do all the languages yourself.

**Comment by funkenstrahlen** [comment]
I think to fix the "shine through" bug the popover needs to set `isAccessibilityElement` for the view it overlays to `true` and `false` on every minfy/max/close action.

The only way to do this I can imagine right now is sending a notification through NSNotificationCenter, so all views can react to it. It might be very hard to find out which view is currently covered by the popup as this can change while navigating through the app with the popup minified.

**Comment by LeoNatan** [comment]
Documentation for `isAccessibilityElement`:

> Assistive applications can get information only about objects that are represented by accessibility elements. Therefore, if you implement a custom control or view that should be accessible to users with disabilities, set this property to YES. **The only exception to this practice is a view that merely serves as a container for other items that should be accessible. Such a view should implement the UIAccessibilityContainer protocol and set this property to NO.**

I have to investigate accessibility for this framework. I have to admit, this is not something I am familiar with.

**Comment by LeoNatan** [comment]
I am also seeing that accessibility notifications should be posted when the screen changes.

```objc
UIAccessibilityPostNotification(UIAccessibilityScreenChangedNotification, ...);
```

Looks like some major changes are required.

**Comment by LeoNatan** [comment]
I fixed the "shine" problem with `accessibilityViewIsModal` on the popup content view. Working on other improvements.

**Comment by funkenstrahlen** [comment]
> I fixed the "shine" problem with accessibilityViewIsModal on the popup content view. Working on other improvements.

That's amazing improvement! Thank you for working on this.

I already have two people waiting to test the voiceover support. This way we can get the best feedback on how good the popup works.

**Comment by LeoNatan** [comment]
I have implemented more accessibility features. The close button will have an accessibility label and a hint. The popup item will support accessibility label and hint, as well as a progress accessibility label. The center of the popup bar will have a button trait and will be customizable. Good things coming soon.

**Comment by LeoNatan** [comment]
Will update the readme with everything related to accessibility, as well as the demo project.

**Comment by funkenstrahlen** [comment]
thanks a lot ❤️

**Comment by jordibruin** [comment]
Very nice to read this conversation between you two. Great work!

**Comment by LeoNatan** [comment]
Done. Please test and see if all you need is supported.

**Comment by funkenstrahlen** [comment]
I just updated my code with your changes and as far as I can see it works great now!

I will have it tested by two voiceover users in the coming weeks. If they suggest any further improvement I'll let you know.

**Comment by LeoNatan** [comment]
May I ask what your app is?

**Comment by funkenstrahlen** [comment]
Sure: https://xenimapp.stefantrauth.de/

It is not released yet. Will be soon. Accessibility support will be added to the released version later as version 1.2.
